File: masthead-api/Masthead.Api/Objects/GraphQLTypes/WeaponFull.cs
using System.Collections.Generic;
using DbgCensus.Core.Objects;

namespace Masthead.Api.Objects.GraphQLTypes;

public record WeaponFull
(
    int ItemId,
    int WeaponId,
    int ItemCategoryId,
    FactionDefinition Faction,
    string? Name,
    string? Description,
    string? ImagePath,
    bool EquipNeedsAmmo,
    short EquipMs,
    short UnequipMs,
    short ToIronSightsMs,
    short FromIronSightsMs,
    short SprintRecoveryMs,
    int NextUseDelayMs,
    decimal TurnModifier,
    decimal MoveModifier,
    short? HeatBleedOffRate,
    short? HeatOverheatPenaltyMs,
    decimal? MeleeDetectWidth,
    decimal? MeleeDetectHeight,
    bool RequiresAmmo,
    int UseCooldownMs,
    IEnumerable<WeaponFull.AmmoSlot> AmmoSlots
) : Weapon(ItemId, WeaponId, ItemCategoryId, Faction, Name, Description, ImagePath)
{
    public record AmmoSlot
    (
        int WeaponSlotIndex,
        short ClipSize,
        short Capacity
    );
}

File: masthead-app/lib/model/weapon_full.dart
class WeaponFull {
  final int itemId;
  final String? name;
  final String? description;
  final String? imagePath;
  final int equipMs;
  final int unequipMs;
  final List<AmmoSlot> ammoSlots;

  WeaponFull(
    this.itemId,
    this.name,
    this.description,
    this.imagePath,
    this.equipMs,
    this.unequipMs,
    this.ammoSlots,
  );

  factory WeaponFull.fromJson(dynamic data) {
    return WeaponFull(
      data['itemId'],
      data['name'],
      data['description'],
      data['imagePath'],
      data['equipMs'],
      data['unequipMs'],
      (data['ammoSlots'] as List).map((e) => AmmoSlot.fromJson(e)).toList(),
    );
  }
}

class AmmoSlot {
  final int capacity;
  final int clipSize;

  AmmoSlot(this.capacity, this.clipSize);

  factory AmmoSlot.fromJson(dynamic data) {
    return AmmoSlot(data['capacity'], data['clipSize']);
  }
}
